pkg/common/utils/utils.go

```diff
@@ -241,3 +243,46 @@ func QueryCapacityForBlockVolumeSnapshotUtil(ctx context.Context, m cnsvolume.Ma
 
 	return snapshotSizeInMb, nil
 }
+
+// Get the datastore reference by datastore URL from a list of datastore references.
+// If the datastore with dsURL can be found in the same datacenter as the given VC
+// and it is also found in the given datastoreList, return the reference of the datastore.
+// Otherwise, return error.
+func GetDatastoreRefByURLFromGivenDatastoreList(
+	ctx context.Context, vc *cnsvsphere.VirtualCenter, datastoreList []types.ManagedObjectReference, dsURL string) (
+	*types.ManagedObjectReference, error) {
+	log := logger.GetLogger(ctx)
+	// get all datacenters in the virtualcenter
+	datacenters, err := vc.GetDatacenters(ctx)
+	if err != nil {
+		log.Errorf("failed to find datacenters from VC: %q, Error: %+v", vc.Config.Host, err)
+		return nil, err
+	}
+	var candidateDsObj *cnsvsphere.Datastore
+	// traverse each datacenter and find the datastore with the specified dsURL
```

Comment:
Can you check if the input dsURL is provided before calling GetDatastoreByURL?

Reply:
We don't need to. The existing code can handles invalid dsURL input, including empty string, very well. I added unit test cases for that.